orderer/consensus/smartbft/chain_test.go

```diff
@@ -0,0 +1,709 @@
+				protoutil.MarshalOrPanic(&cb.Metadata{
+					Value: protoutil.MarshalOrPanic(&cb.OrdererBlockMetadata{
+						LastConfig: &cb.LastConfig{
+							// The last config index is the genesis blockZ
+							Index: 0,
+						},
+						ConsenterMetadata: nil,
+					}),
+					Signatures: nil,
+				}),
+				// BlockMetadataIndex_LAST_CONFIG
+				{},
+				// BlockMetadataIndex_TRANSACTIONS_FILTER
+				{},
+				// BlockMetadataIndex_ORDERER
+				{},
+				// BlockMetadataIndex_COMMIT_HASH
+				{},
+			},
+		},
+	}
+}
+
+func generateNonce() []byte {
+	nonceBuf := new(bytes.Buffer)
+	err := binary.Write(nonceBuf, binary.LittleEndian, nonce)
+	if err != nil {
+		panic("Cannot generate nonce")
+	}
+	nonce++
+	return nonceBuf.Bytes()
+}
+
+func createAllGoodPolicy(t *testing.T) policies.Policy {
+	policyMock := policiesMocks.NewPolicy(t)
+	policyMock.EXPECT().EvaluateIdentities(mock.Anything).Return(nil).Maybe()
+	policyMock.EXPECT().EvaluateSignedData(mock.Anything).Return(nil).Maybe()
+	return policyMock
+}
+
+func createEndorserTxEnvelope(message string, channelId string) *cb.Envelope {
+	return &cb.Envelope{
+		Payload: protoutil.MarshalOrPanic(&cb.Payload{
+			Header: &cb.Header{
+				ChannelHeader: protoutil.MarshalOrPanic(&cb.ChannelHeader{
+					Type:      int32(cb.HeaderType_ENDORSER_TRANSACTION),
+					ChannelId: channelId,
+				}),
+				SignatureHeader: protoutil.MarshalOrPanic(&cb.SignatureHeader{
+					Creator: protoutil.MarshalOrPanic(&msp.SerializedIdentity{
+						Mspid:   "mockMSP",
+						IdBytes: []byte("mockClient"),
+					}),
+					Nonce: generateNonce(),
+				}),
+			},
+			Data: []byte(message),
+		}),
+	}
+}
+
+func updateRuntimeConfigWithBlock(rtc *smartbft.RuntimeConfig, lastBlock *cb.Block) {
+	rtc.LastBlock = lastBlock
+	rtc.LastCommittedBlockHash = hex.EncodeToString(protoutil.BlockHeaderHash(lastBlock.Header))
+	if protoutil.IsConfigBlock(lastBlock) {
+		rtc.LastConfigBlock = lastBlock
+	}
+}
+
+func maxTimeConfigFactory(maxTime time.Duration) ChainConfigFactory {
+	return func(nodeId uint64) *types.Configuration {
+		config := defaultConfigFactory(nodeId)
+		config.RequestForwardTimeout = maxTime
+		config.RequestComplainTimeout = maxTime
+		config.RequestAutoRemoveTimeout = maxTime
+		config.ViewChangeResendInterval = maxTime
+		config.ViewChangeTimeout = maxTime
+		config.LeaderHeartbeatTimeout = maxTime
+		config.CollectTimeout = maxTime
+		return config
+	}
+}
+
+func allSameIds(ids []uint64) bool {
+	if len(ids) == 0 {
+		return true
+	}
+	for _, id := range ids {
+		if id != ids[0] {
+			return false
+		}
+	}
+	return true
+}
+
+//////////////////////////////////////////////////////////////
+// TestNetwork
+//////////////////////////////////////////////////////////////
+
+type TestNetwork struct {
+	t                *testing.T
+	g                *WithT
+	clusterState     *ClusterState
+	ChannelId        string
+	nodeIdToChain    map[uint64]*ChainUnderTest
+	nodeIdToActivity map[uint64]bool
+	lock             sync.Mutex
+}
```

Comment:
Since it is quite big, move the entire testing infrastructure to a separate file (util_network_test.go, etc), such that in this file we only have (mostly) test cases.